storage_controller/src/tenant_shard.rs

```diff
@@ -646,6 +646,44 @@ impl TenantShard {
         Ok(())
     }
 
+    /// Reschedule this tenant shard to one of its secondary locations. Returns a scheduling error
+    /// if the swap is not possible and leaves the intent state in its original state.
+    ///
+    /// Arguments:
+    /// `attached_to`: the currently attached location matching the intent state (may be None if the
+    /// shard is not attached)
+    /// `promote_to`: an optional secondary location of this tenant shard. If set to None, we ask
+    /// the scheduler to recommend a node
+    pub(crate) fn reschedule_to_secondary(
+        &mut self,
+        attached_to: Option<NodeId>,
+        promote_to: Option<NodeId>,
+        scheduler: &mut Scheduler,
+    ) -> Result<(), ScheduleError> {
+        let promote_to = match promote_to {
+            Some(node) => node,
+            None => match scheduler.node_preferred(self.intent.get_secondary()) {
+                Some(node) => node,
+                None => {
+                    return Err(ScheduleError::ImpossibleConstraint);
+                }
+            },
+        };
+
+        assert!(self.intent.get_secondary().contains(&promote_to));
+
+        if let Some(node) = attached_to {
+            let demoted = self.intent.demote_attached(scheduler, node);
+            if !demoted {
+                return Err(ScheduleError::ImpossibleConstraint);
+            }
+        }
+
+        self.intent.promote_attached(scheduler, promote_to);
+
```

Comment:
Let's increment self.sequence here.  It will usually work anyway, but there would be an edge case if a the reconcile is already in flight, if we haven't incremented the sequence then a call to maybe_reconcile immediately after calling this would just wait for the existing reconciler (eventually reconciliation would happen in background loop)

Reply:
Thanks: [f2fe1ea](https://github.com/neondatabase/neon/pull/8119/commits/f2fe1ea35542e88d4e22c97887243041bfcdb1b8)